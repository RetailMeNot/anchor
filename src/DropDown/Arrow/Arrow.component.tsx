// VENDOR
import styled, { css } from '@xstyled/styled-components';
import { th } from '@xstyled/system';
// COMPONENTS
import { Position } from '../DropDown.component';

export const DefaultSize = '0.625rem';

export interface ArrowProps {
    position: Position;
    indent?: string;
    size?: string;
    border?: string;
    shadow?: string;
    background?: string;
}

export const Arrow = styled('div')<ArrowProps>`
    position: absolute;
    box-sizing: border-box;
    ${({ background, shadow, border, size = DefaultSize }) =>
        css({
            width: size,
            height: size,
            border,
            background: th.color(background),
            boxShadow: shadow,
        })}
    z-index: -1;
    transform: rotate(45deg);

    ${({ position, indent, size = DefaultSize }) => {
        switch (position) {
            case 'bottom':
                return css({
                    top: `calc(-${size} / 2)`,
                    left: `calc(50% - ${size} / 2)`,
                });
            case 'bottomEnd':
                return css({
                    top: `calc(-${size} / 2)`,
                    right: indent,
                });
            case 'bottomStart':
                return css({
                    top: `calc(-${size} / 2)`,
                    left: indent,
                });
            case 'top':
                return css({
                    bottom: `calc(-${size} / 2)`,
                    right: `calc(50% - ${size} / 2)`,
                });
            case 'topEnd':
                return css({
                    bottom: `calc(-${size} / 2)`,
                    right: indent,
                });
            case 'topStart':
                return css({
                    bottom: `calc(-${size} / 2)`,
                    left: indent,
                });
            case 'right':
                return css({
                    top: `calc(50% - ${size} / 2)`,
                    left: `calc(-${size} / 2)`,
                });
            case 'rightEnd':
                return css({
                    bottom: indent,
                    left: `calc(-${size} / 2)`,
                });
            case 'rightStart':
                return css({
                    top: indent,
                    left: `calc(-${size} / 2)`,
                });
            case 'left':
                return css({
                    top: `calc(50% - ${size} / 2)`,
                    right: `calc(-${size} / 2)`,
                });
            case 'leftEnd':
                return css({
                    bottom: indent,
                    right: `calc(-${size} / 2)`,
                });
            case 'leftStart':
                return css({
                    top: indent,
                    right: `calc(-${size} / 2)`,
                });
        }
        return null;
    }}
`;
