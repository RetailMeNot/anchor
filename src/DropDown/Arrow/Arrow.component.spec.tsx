// REACT
import * as React from 'react';
import * as renderer from 'react-test-renderer';
// COMPONENT
import { renderArrow } from './Arrow.component';
// ENZYME
import { shallow, mount } from 'enzyme';

// TEST SETUP
const subject = (
    <div>
        {renderArrow('top', true)}
        <br />
        {renderArrow('bottom', true)}
        <br />
        {renderArrow('left', true)}
        <br />
        {renderArrow('right', true)}
    </div>
);
const wrapper = mount(subject);
const component = shallow(subject);

describe('Component: DropDown', () => {
    describe('Component: Arrow', () => {
        test('should render in the DOM', () => {
            expect(subject).toBeDefined();
            expect(wrapper).toBeDefined();
            expect(component).toBeDefined();
            const tree = renderer.create(subject).toJSON();

            expect(tree).toMatchSnapshot();
        });
    });
});
