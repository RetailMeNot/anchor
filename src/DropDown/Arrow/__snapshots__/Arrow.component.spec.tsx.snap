// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: DropDown Component: Arrow should render in the DOM 1`] = `
.c0 {
  position: absolute;
  box-sizing: border-box;
  width: 0.625rem;
  height: 0.625rem;
  z-index: -1;
  -webkit-transform: rotate(45deg);
  -ms-transform: rotate(45deg);
  transform: rotate(45deg);
  bottom: calc(-0.625rem / 2);
  right: calc(50% - 0.625rem / 2);
}

.c1 {
  position: absolute;
  box-sizing: border-box;
  width: 0.625rem;
  height: 0.625rem;
  z-index: -1;
  -webkit-transform: rotate(45deg);
  -ms-transform: rotate(45deg);
  transform: rotate(45deg);
  top: calc(-0.625rem / 2);
  left: calc(50% - 0.625rem / 2);
}

.c2 {
  position: absolute;
  box-sizing: border-box;
  width: 0.625rem;
  height: 0.625rem;
  z-index: -1;
  -webkit-transform: rotate(45deg);
  -ms-transform: rotate(45deg);
  transform: rotate(45deg);
  top: calc(50% - 0.625rem / 2);
  right: calc(-0.625rem / 2);
}

.c3 {
  position: absolute;
  box-sizing: border-box;
  width: 0.625rem;
  height: 0.625rem;
  z-index: -1;
  -webkit-transform: rotate(45deg);
  -ms-transform: rotate(45deg);
  transform: rotate(45deg);
  top: calc(50% - 0.625rem / 2);
  left: calc(-0.625rem / 2);
}

<div>
  <div
    className="c0"
  />
  <div
    className="c1"
  />
  <div
    className="c2"
  />
  <div
    className="c3"
  />
</div>
`;
